### PR TITLE
fix(solar.make_chromaticdelay): NaN's in gradient

### PR DESCRIPTION
**Pull request by davecwright3**
I found `NaN`'s coming from this function in the gradients of both single-pulsar and full-PTA likelihoods. For very large negative numbers in the exponent, the gradient wrt the timescale $$\tau$$ is  

```math
\propto \frac{-\Delta t}{\tau}  e^{\tfrac{-\Delta t}{\tau}},
```

where $\Delta t$ is the time after the start of the event.  $\Delta t$ can be very large, which creates a product of a very large negative number and a number very close to zero. This produces `NaN`'s (I double checked I was working in double precision).

My first attempt at fixing this was to rewrite the dip in the form

```math
10^{\log_{10}(A)} e^{-\Delta t / 10^{\tau}} f^{-\gamma} =  \textrm{exp}\left(\log_{10}(A)\ln{(10)} + -\Delta t / 10^{\tau}  + -\gamma\ln{f}\right ),
```

which I hoped would stabilize the computation. It also *can* push the exponent towards more positive values for $f <$ 1400 MHz. It fixes some cases, but $$\Delta t$$ can still become large enough to make `NaN`s. I then tried a few things like clipping, scaling, etc., but the solution that worked ended up being the most simple one.

I added a `jnp.where` to filter `NaN`'s out of the exponent. If someone has a more intelligent solution to this issue, I'd be happy to change this PR.

There's also an updated docstring included in this PR.

### CODE REVIEW

File: src/discovery/solar.py [hunk 1]
```diff
@@ -45,13 +45,95 @@ def chromaticdelay(toas, freqs, t0, log10_Amp, log10_tau, idx):
...
+            ln_10 * log10_Amp - dt / (10**log10_tau) + idx * ln_invnormfreqs,
+            0.0,
+        )
+
+        # Filter out any `NaN`s. A few things were tried here (e.g. clipping), but
+        # they didn't handle the `NaN`s in the gradient. This, while brute force,
+        # does the trick.
+        vals = matrix.jnp.where(matrix.jnp.isnan(vals), 0.0, vals)
+
+        return matrix.jnp.where(dt_mask, -1.0 * matrix.jnp.exp(vals), vals)
```
Comment (davecwright3) [inline src/discovery/solar.py:136]: I tried putting the NaN mask *after* the `jnp.exp`, but it still produced NaN's in the gradient. Putting the mask on `vals` removed the NaN's.

**Comment by davecwright3** [comment]
Here's a minimal reproducible example 

```python
>>> j1713 = ds.Pulsar.read_feather("./data/v1p1_de440_pint_bipm2019-J1713+0747.feather")
>>> j1713_likelihood =  ds.PulsarLikelihood(
    [
        j1713.residuals,
        ds.makenoise_measurement(j1713, j1713.noisedict),
        ds.makegp_ecorr(j1713, j1713.noisedict),
        ds.makegp_timing(j1713, svd=True),
        ds.makedelay(j1713, ds.make_solardm(j1713), name="sw", common=["sw_n_earth",]),
        ds.makedelay( j1713, ds.make_chromaticdelay(j1713), name="dmexp_0" ),
        ds.makedelay( j1713, ds.make_chromaticdelay(j1713), name="dmexp_1")
    ]
)
>>> j1713_params = {
    'J1713+0747_dmexp_0_idx': 3.0,
    'J1713+0747_dmexp_0_log10_Amp': -6.0,
    'J1713+0747_dmexp_0_log10_tau': 2.0,
    'J1713+0747_dmexp_0_t0': 54749.0,
    'J1713+0747_dmexp_1_idx': 1.5,
    'J1713+0747_dmexp_1_log10_Amp': -6.0,
    'J1713+0747_dmexp_1_log10_tau': 1.5,
    'J1713+0747_dmexp_1_t0': 57509.0,
    'sw_n_earth': 7.5,
}


>>> jax.grad(j1713_likelihood.logL)(j1713_params)

{'J1713+0747_dmexp_0_idx': Array(14.00639279, dtype=float64, weak_type=True),
 'J1713+0747_dmexp_0_log10_Amp': Array(22.62681313, dtype=float64, weak_type=True),
 'J1713+0747_dmexp_0_log10_tau': Array(35.44668872, dtype=float64, weak_type=True),
 'J1713+0747_dmexp_0_t0': Array(0.098267, dtype=float64, weak_type=True),
 'J1713+0747_dmexp_1_idx': Array(-89.88917482, dtype=float64, weak_type=True),
 'J1713+0747_dmexp_1_log10_Amp': Array(101.46822264, dtype=float64, weak_type=True),
 'J1713+0747_dmexp_1_log10_tau': Array(94.28126695, dtype=float64, weak_type=True),
 'J1713+0747_dmexp_1_t0': Array(1.39352372, dtype=float64, weak_type=True),
 'sw_n_earth': Array(-0.58970052, dtype=float64, weak_type=True)}

>>> jax.grad(j1713_likelihood.logL)(j1713_params | {'J1713+0747_dmexp_1_log10_tau':0.7 })

{'J1713+0747_dmexp_0_idx': Array(13.79672801, dtype=float64, weak_type=True),
 'J1713+0747_dmexp_0_log10_Amp': Array(22.28019812, dtype=float64, weak_type=True),
 'J1713+0747_dmexp_0_log10_tau': Array(35.3774397, dtype=float64, weak_type=True),
 'J1713+0747_dmexp_0_t0': Array(0.09676167, dtype=float64, weak_type=True),
 'J1713+0747_dmexp_1_idx': Array(nan, dtype=float64, weak_type=True),
 'J1713+0747_dmexp_1_log10_Amp': Array(nan, dtype=float64, weak_type=True),
 'J1713+0747_dmexp_1_log10_tau': Array(nan, dtype=float64, weak_type=True),
 'J1713+0747_dmexp_1_t0': Array(nan, dtype=float64, weak_type=True),
 'sw_n_earth': Array(-0.58058305, dtype=float64, weak_type=True)}

>>> jax.grad(j1713_likelihood.logL)(j1713_params | {'J1713+0747_dmexp_1_log10_tau':0.8 })

{'J1713+0747_dmexp_0_idx': Array(13.81124307, dtype=float64, weak_type=True),
 'J1713+0747_dmexp_0_log10_Amp': Array(22.30246742, dtype=float64, weak_type=True),
 'J1713+0747_dmexp_0_log10_tau': Array(35.37423915, dtype=float64, weak_type=True),
 'J1713+0747_dmexp_0_t0': Array(0.09685839, dtype=float64, weak_type=True),
 'J1713+0747_dmexp_1_idx': Array(-20.87471305, dtype=float64, weak_type=True),
 'J1713+0747_dmexp_1_log10_Amp': Array(23.79458728, dtype=float64, weak_type=True),
 'J1713+0747_dmexp_1_log10_tau': Array(35.90039461, dtype=float64, weak_type=True),
 'J1713+0747_dmexp_1_t0': Array(1.63780611, dtype=float64, weak_type=True),
 'sw_n_earth': Array(-0.58029096, dtype=float64, weak_type=True)}
```

For some reason, 0.8 is the limit below which the gradients become NaN's.

**Comment by davecwright3** [comment]
The last call to `jnp.where` for the `NaN` mask is not needed. I made the usual mistake of turning two knobs at once, and it turns out that calculating the exponent outside of the call to `jnp.exp` is what fixes the issues. I'd bet that it's some issue with the different `jaxpr`'s that JAX's autodiff creates, but I haven't inspected it in detail.

**Comment by davecwright3** [comment]
Apologies for the back on forth, I've finally found the root issue. `JAX`'s reverse-mode autodiff still evaluates the gradient at the `False` positions for the expression in a `jnp.where` https://docs.jax.dev/en/latest/faq.html#gradients-contain-nan-where-using-where. For our case, this means we evaluate at all `-dt`. For `dt` negative enough, the expression `jnp.exp(-dt / 10**log10_tau` overflows and returns an `inf`, giving a `NaN` gradient.

This explains why ~0.8 was the magic number that produced `NaN`'s.

```python
>>> j1713 = ds.Pulsar.read_feather("./data/v1p1_de440_pint_bipm2019-J1713+0747.feather")
>>> j1713_params = {
    'J1713+0747_dmexp_1_idx': 1.5,
    'J1713+0747_dmexp_1_log10_Amp': -6.0,
    'J1713+0747_dmexp_1_log10_tau': 1.5,
    'J1713+0747_dmexp_1_t0': 57509.0,
}
>>> toadays = jnp.array(j1713.toas / ds.const.day)
>>> dt = toadays - j1713_params["J1713+0747_dmexp_1_t0"]
>>> dt.min(), dt.max()
(Array(-4115.44121685, dtype=float64), Array(1562.00791291, dtype=float64))

>>> tau = 10** 0.8
>>> jnp.exp(-dt.min() / tau)
Array(1.86245505e+283, dtype=float64)

>>> tau = 10** 0.7633
>>> jnp.exp(-dt.min() / tau)
Array(1.77140153e+308, dtype=float64)

>>> tau = 10** 0.76
>>> jnp.exp(-dt.min() / tau)
Array(inf, dtype=float64)
```
Which of course makes perfect sense because we've reached the maximum float that can be represented with 64 bits

```python
>>> import sys
>>> sys.float_info.max
1.7976931348623157e+308 
```

### The solution
JAX docs (and other autodiff resources) recommend the "double where" trick, which I accidentally implemented. The first `jnp.where` in this PR keeps the values that can overflow the exponential from ever entering it. A quick solution I came up with is to move the `jnp.where` inside the exponential and return a `-inf` for `dt < 0.0`, but in my bench-marking this was actually ~100us slower than what I had already written.
